common/exception/TypeDBClientException.java

```diff
@@ -98,6 +109,25 @@ private static boolean isUnimplementedMethod(StatusRuntimeException statusRuntim
         return statusRuntimeException.getStatus().getCode() == Status.Code.UNIMPLEMENTED;
     }
 
+    private static TypeDBClientException tryGetRstStreamCause(StatusRuntimeException sre) {
+        if (sre.getStatus().getCode() == Status.Code.UNAVAILABLE) {
+            if (sre.getCause() instanceof javax.net.ssl.SSLHandshakeException) {
+                if (sre.getCause().getMessage().contains("PKIX path building failed")) {
+                    return new TypeDBClientException(sre.getCause(), CLUSTER_SSL_CERTIFICATE_NOT_VALIDATED);
+                } else if (sre.getCause().getMessage().contains("No subject alternative DNS name")) {
+                    return new TypeDBClientException(sre.getCause(), CLUSTER_SSL_CERTIFICATE_INVALID_FOR_HOSTNAME);
+                } else return new TypeDBClientException(sre.getCause(), CLUSTER_SSL_HANDSHAKE_FAILED, sre.getMessage());
+            } else if (sre.getCause() instanceof io.netty.handler.ssl.NotSslRecordException) {
+                return new TypeDBClientException(sre.getCause(), CLUSTER_SERVER_NOT_ENCRYPTED);
+            } else if (sre.getStatus().getDescription() != null &&
+                    sre.getStatus().getDescription().contains("Network closed for unknown reason")) {
+                // Too generic a reason to throw. Wrap it in UNABLE_TO_CONNECT so the client tries other nodes.
+                return new TypeDBClientException(new TypeDBClientException(sre.getCause(), CLUSTER_CONNECTION_CLOSED_UKNOWN), UNABLE_TO_CONNECT);
```

Comment:
It's awkward to have an exception in an exception, can we directly use `CLOSED_UNKNOWN` alone?

Reply:
I think this one you'd need to take up with Ganesh while im away or wait till next week

Reply:
Flagging to revisit